src/ripple/app/tx/impl/PayChan.cpp

```diff
@@ -0,0 +1,470 @@
+        SourceTag (optional)
+
+    PaymentChannelFund
+
+        Add additional funds to the payment channel. Only the channel owner may
+        use this transaction. The parameters are:
+        Channel
+            The 256-bit ID of the channel.
+        Amount
+            The amount of XRP to add.
+        Expiration (optional)
+            Time the channel closes. The transaction will fail if the expiration
+            times does not satisfy the SettleDelay constraints.
+
+    PaymentChannelClaim
+
+        Place a claim against an existing channel. The parameters are:
+        Channel
+            The 256-bit ID of the channel.
+        Balance (optional)
+            The total amount of XRP delivered after this claim is processed (optional, not
+            needed if just closing).
+        Amount (optional)
+            The amount of XRP the signature is for (not needed if equal to Balance or just
+            closing the line).
+        Signature (optional)
+            Authorization for the balance above, signed by the owner (optional,
+            not needed if closing or owner is performing the transaction). The
+            signature if for the following message: CLM\0 followed by the
+            256-bit channel ID, and a 64-bit integer drops.
+        PublicKey (optional)
+           The public key that made the signature (optional, required if a signature
+           is present)
+        Flags
+            tfCloseChannel
+                Request that the channel be closed
+            tfRenewChannel
+                Request that the channel's expiration be reset. Only the owner
+                may renew a channel.
+
+*/
+
+//------------------------------------------------------------------------------
+
+static
+TER
+closeChannel (
+    std::shared_ptr<SLE> const& slep,
+    ApplyView& view,
+    uint256 const& key,
+    beast::Journal j)
+{
+    AccountID const src = (*slep)[sfAccount];
+    // Remove PayChan from owner directory
+    {
+        auto const page = (*slep)[sfOwnerNode];
+        TER const ter = dirDelete (view, true, page, keylet::ownerDir (src).key,
+            key, false, page == 0, j);
+        if (!isTesSuccess (ter))
+            return ter;
+    }
+
+    // Transfer amount back to owner, decrement owner count
+    auto const sle = view.peek (keylet::account (src));
+    if (!sle)
+        return terNO_ACCOUNT;
+    assert ((*slep)[sfAmount] >= (*slep)[sfBalance]);
+    (*sle)[sfBalance] =
+        (*sle)[sfBalance] + (*slep)[sfAmount] - (*slep)[sfBalance];
+    (*sle)[sfOwnerCount] = (*sle)[sfOwnerCount] - 1;
+    view.update (sle);
+
+    // Remove PayChan from ledger
+    view.erase (slep);
+    return tesSUCCESS;
+}
+
+//------------------------------------------------------------------------------
+
+TER
+PayChanCreate::preflight (PreflightContext const& ctx)
+{
+    if (!ctx.rules.enabled (featurePayChan, ctx.app.config ().features))
```

Comment:
We seem to be allowing negative values through to sfSettleDelay. Since its unsigned, we might end up with an erroneous value.